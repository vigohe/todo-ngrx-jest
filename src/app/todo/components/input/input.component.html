<div>
  <input #todo_input placeholder="Add a todo" (keyup.enter)="addTodo(todo_input.value); todo_input.value=''" />
</div>
