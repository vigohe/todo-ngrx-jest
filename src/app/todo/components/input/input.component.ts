import { Component, OnInit, Output, EventEmitter } from '@angular/core';

@Component({
  selector: 'app-input',
  templateUrl: './input.component.html',
  styleUrls: ['./input.component.css']
})
export class InputComponent implements OnInit {

  @Output() add = new EventEmitter<String>();

  constructor() { }

  ngOnInit() {
  }

  addTodo(todoText: String) {
    this.add.emit(todoText);
  }

}
