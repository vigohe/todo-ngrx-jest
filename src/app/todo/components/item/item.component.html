<h1 [style.text-decoration]="todo.complete ? 'line-through' : ''">
    {{todo.text}} <input type="checkbox" (change)="onChecked(todo)" [checked]="todo.complete"/> <button (click)="onRemove(todo)">X</button>
</h1>
