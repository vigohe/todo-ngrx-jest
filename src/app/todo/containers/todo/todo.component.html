<app-input (add)="addTodo($event)"></app-input>
<app-list [todos]="todos | async" (toggle)="onToggle($event)" (remove)="onRemove($event)"></app-list>
