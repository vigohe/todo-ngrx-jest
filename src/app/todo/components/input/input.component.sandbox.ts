import { sandboxOf } from 'angular-playground';
import { InputComponent } from './input.component';

export default sandboxOf(InputComponent)
  .add('with simple text', {
    template: `<app-input></app-input>`
  });
