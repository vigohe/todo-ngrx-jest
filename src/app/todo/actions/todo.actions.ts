import { Action } from '@ngrx/store';
import { Todo } from '../models/todo';

export enum TodoActionTypes {
  AddTodo = '[Todo] ADD',
  RemoveTodo = '[Todo] Remove',
  ToggleTodo = '[Todo] Toggle'
}

export class AddTodo implements Action {
  readonly type = TodoActionTypes.AddTodo;
  constructor(public payload : string) {};
}

export class RemoveTodo implements Action {
  readonly type = TodoActionTypes.RemoveTodo;
  constructor(public payload : Todo) {};
}

export class ToggleTodo implements Action {
  readonly type = TodoActionTypes.ToggleTodo;
  constructor(public payload : Todo) {};
}


export type TodoActions = AddTodo | RemoveTodo | ToggleTodo;
