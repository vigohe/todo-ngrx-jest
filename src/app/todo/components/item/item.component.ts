import { Component, OnInit, Input, Output, EventEmitter } from '@angular/core';
import { Todo } from '../../models/todo';

@Component({
  selector: 'app-item',
  templateUrl: './item.component.html',
  styleUrls: ['./item.component.css']
})
export class ItemComponent implements OnInit {

  @Input() todo: Todo;
  @Output() toggle = new EventEmitter<Todo>();
  @Output() remove = new EventEmitter<Todo>();

  constructor() { }

  ngOnInit() {
  }

  onChecked(todo: Todo) {
    this.toggle.emit(todo);
  }

  onRemove(todo: Todo) {
    this.remove.emit(todo);
  }

}
