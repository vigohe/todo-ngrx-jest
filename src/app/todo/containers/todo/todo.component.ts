import { Component, OnInit } from '@angular/core';
import { Store } from '@ngrx/store';
import * as fromStore from '../../reducers/todo.reducer';
import { AddTodo, ToggleTodo, RemoveTodo } from '../../actions/todo.actions';
import { Todo } from '../../models/todo';
import { Observable } from 'rxjs';
import { State } from '../../../reducers';

@Component({
  selector: 'app-todo',
  templateUrl: './todo.component.html',
  styleUrls: ['./todo.component.css']
})
export class TodoComponent implements OnInit {
  public todos: Observable<Todo[]>;

  constructor(private store: Store<State>) { }

  ngOnInit() {
    this.todos = this.store.select((state: State) => state.todo.entities );
  }

  addTodo(newTodo: string) {
    this.store.dispatch(new AddTodo(newTodo));
  }

  onToggle(todo: Todo) {
    this.store.dispatch(new ToggleTodo(todo));
  }

  onRemove(todo: Todo) {
    this.store.dispatch(new RemoveTodo(todo));
  }
}
