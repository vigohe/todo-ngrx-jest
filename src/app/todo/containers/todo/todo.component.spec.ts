import { ComponentFixture, TestBed } from '@angular/core/testing';

import { TodoComponent } from './todo.component';
import { Store, StoreModule } from '@ngrx/store';
import { InputComponent } from '../../components/input/input.component';
import { ListComponent } from '../../components/list/list.component';
import { AddTodo, ToggleTodo, RemoveTodo } from '../../actions/todo.actions';
import { ItemComponent } from '../../components/item/item.component';
import { reducers, metaReducers } from '../../../reducers';
import { Todo } from '../../models/todo';

describe('TodoComponent', () => {
  let component: TodoComponent;
  let fixture: ComponentFixture<TodoComponent>;
  let store: Store<any>;

  beforeEach(async() => {
    TestBed.configureTestingModule({
      imports: [ StoreModule.forRoot(reducers, { metaReducers }) ],
      declarations: [ TodoComponent, ListComponent, InputComponent, ItemComponent]
    });

    await TestBed.compileComponents();
  });

  beforeEach(() => {
    fixture = TestBed.createComponent(TodoComponent);
    component = fixture.componentInstance;
    store = TestBed.get(Store);

    spyOn(store, 'dispatch').and.callThrough();
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should compile', () => {
    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });


  it('should dispatch a todo.Add action on input', () => {
    const $event: string = 'GO TO SLEEP';
    const action = new AddTodo($event);

    component.addTodo($event);

    expect(store.dispatch).toHaveBeenCalledWith(action);
  });

  it('should dispatch a todo.Toggle action onToggle', () => {
    const $event: Todo = {
      id: 1,
      text: 'asdf',
      complete: true
    };
    const action = new ToggleTodo($event);
    component.onToggle($event);
    expect(store.dispatch).toHaveBeenCalledWith(action);
  });

  it('should dispatch a todo.Remove action onRemove', () => {
    const $event: Todo = {
      id: 1,
      text: 'asdf',
      complete: true
    };
    const action = new RemoveTodo($event);
    component.onRemove($event);
    expect(store.dispatch).toHaveBeenCalledWith(action);
  });

});
