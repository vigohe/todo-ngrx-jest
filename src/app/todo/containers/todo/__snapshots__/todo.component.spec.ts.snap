// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoComponent should compile 1`] = `
<app-todo
  store={[Function Store]}
  todos={[Function Store]}
>
  <app-input>
    <div>
      
    
      <input
        placeholder="Add a todo"
      />
      
  
    </div>
    
  
  </app-input>
  <app-list
    ng-reflect-todos=""
  >
    
    
  
  </app-list>
  
</app-todo>
`;
